stackslib/src/net/inv/nakamoto.rs

```diff
@@ -0,0 +1,215 @@
+                parent_tenure_id_consensus_hash: tenure.prev_tenure_id_consensus_hash,
+            }),
+        )
+    }
+}
+
+/// This struct represents cached inventory data loaded from Nakamoto headers.
+/// It is of the utmost importance that inventory message generation is _fast_, and incurs as
+/// little I/O overhead as possible, given how essential these messages are to nodes trying to keep
+/// in sync.  By caching (immutable) tenure data in this struct, we can enusre that this happens
+/// all the time except for during node bootup.
+pub struct InvGenerator {
+    processed_tenures: HashMap<ConsensusHash, Option<InvTenureInfo>>,
+    sortitions: HashMap<ConsensusHash, InvSortitionInfo>,
+}
+
+impl InvGenerator {
+    pub fn new() -> Self {
+        Self {
+            processed_tenures: HashMap::new(),
+            sortitions: HashMap::new(),
+        }
+    }
+
+    /// Get a processed tenure. If it's not cached, then load it.
+    /// Returns Some(..) if there existed a tenure-change tx for this given consensus hash
+    fn get_processed_tenure(
+        &mut self,
+        chainstate: &StacksChainState,
+        tenure_id_consensus_hash: &ConsensusHash,
+    ) -> Result<Option<InvTenureInfo>, NetError> {
+        let cur_tenure_info_opt =
+            if let Some(info_opt) = self.processed_tenures.get(&tenure_id_consensus_hash) {
+                Ok((*info_opt).clone())
+            } else {
+                let loaded_info_opt = InvTenureInfo::load(chainstate, &tenure_id_consensus_hash)?;
+                self.processed_tenures
+                    .insert(tenure_id_consensus_hash.clone(), loaded_info_opt.clone());
+                Ok(loaded_info_opt)
+            };
+        cur_tenure_info_opt
+    }
+
+    /// Generate an block inventory bit vector for a reward cycle.
+    /// The bit vector is "big-endian" -- the first bit is the oldest sortition, and the last bit is
+    /// the newest sortition.  It is structured as follows:
+    /// * Bit 0 is the sortition at the start of the given reward cycle
+    /// * Bit i is 1 if there was a tenure-start for the ith sortition in the reward cycle, and 0
+    /// if not.
+    ///
+    /// Populate the cached data lazily.
+    ///
+    /// * `tip` is the canonical sortition tip
+    /// * `chainstate` is a handle to the chainstate DB
+    /// * `reward_cycle` is the reward cycle for which to generate the inventory
+    ///
+    /// The resulting bitvector will be truncated if `reward_cycle` is the current reward cycle.
+    pub fn make_tenure_bitvector(
+        &mut self,
+        tip: &BlockSnapshot,
+        sortdb: &SortitionDB,
+        chainstate: &StacksChainState,
+        reward_cycle: u64,
+    ) -> Result<Vec<bool>, NetError> {
+        let ih = sortdb.index_handle(&tip.sortition_id);
+        let reward_cycle_end_height = sortdb
+            .pox_constants
+            .reward_cycle_to_block_height(sortdb.first_block_height, reward_cycle + 1)
```

Comment:
Wouldn't the end height of the reward cycle be -1? D:

Reply:
Sure -- `reward_cycle_to_block_height()` returns a value that is `mod 1` of the reward cycle length.  The `- 2` gets us the last block in the last reward cycle. So if `reward_cycle_to_block_height` returns (for example) 823551, then this would evaluate to 8235549.  This is the bitcoin block height we actually want to start with, because the loop invariant will scan down to block 821450 (inclusive) due to the fact that `block_height_to_reward_cycle` will return the same reward cycle for 823450 as 823551.

Reply:
ahhh okay this makes sense.